### ISSUE TITLE
snipMate, coc-snippets, Vimwiki <Tab>, <S-Tab>, <CR> key issue

### ISSUE THREAD

**Issue by N-Tek** [vimwiki/vimwiki#1093]
Prior to submitting a new issue make sure to complete these steps:

- [x] Checkout the `dev` branch and confirm the issue is present there as well.
      The `dev` branch contains fixes that may not have been merged to `master` yet.

- For `master` branch following issue related lines are used in `.vimrc` for installation with `vim-plug`

      call plug#begin('~/.vim/plugged')
      ...
      Plug 'neoclide/coc.nvim', {'branch': 'release'}
      let g:coc_global_extensions = ['coc-bookmark',
	                                             \ 'coc-calc',
			                             \ 'coc-clangd',
			                             \ 'coc-cmake',
			                             \ 'coc-css',
			                             \ 'coc-cssmodules',
			                             \ 'coc-git',
			                             \ 'coc-highlight',
			                             \ 'coc-html',
			                             \ 'coc-json',
			                             \ 'coc-markdownlint',
			                             \ 'coc-r-lsp',
			                             \ 'coc-sh',
			                             \ 'coc-snippets',
			                             \ 'coc-stylelint',
			                             \ 'coc-stylelintplus',
			                             \ 'coc-svg',
			                             \ 'coc-template',
			                             \ 'coc-todolist',
			                             \ 'coc-tsserver',
			                             \ 'coc-vimlsp',
			                             \ 'coc-xml']
      ...
      Plug 'tomtom/tlib_vim'
      Plug 'MarcWeber/vim-addon-mw-utils'
      Plug  'garbas/vim-snipmate'

      Plug 'honza/vim-snippets'
      ...
      Plug 'vimwiki/vimwiki'
      call plug#end()

- For `dev` branch following lines are used in `.vimrc` for installation with `vim-plug`

      call plug#begin('~/.vim/plugged')
      ...
      Plug 'neoclide/coc.nvim', {'branch': 'release'}
      let g:coc_global_extensions = ['coc-bookmark',
	                                             \ 'coc-calc',
			                             \ 'coc-clangd',
			                             \ 'coc-cmake',
			                             \ 'coc-css',
			                             \ 'coc-cssmodules',
			                             \ 'coc-git',
			                             \ 'coc-highlight',
			                             \ 'coc-html',
			                             \ 'coc-json',
			                             \ 'coc-markdownlint',
			                             \ 'coc-r-lsp',
			                             \ 'coc-sh',
			                             \ 'coc-snippets',
			                             \ 'coc-stylelint',
			                             \ 'coc-stylelintplus',
			                             \ 'coc-svg',
			                             \ 'coc-template',
			                             \ 'coc-todolist',
			                             \ 'coc-tsserver',
			                             \ 'coc-vimlsp',
			                             \ 'coc-xml']
      ...
      Plug 'tomtom/tlib_vim'
      Plug 'MarcWeber/vim-addon-mw-utils'
      Plug  'garbas/vim-snipmate'

      Plug 'honza/vim-snippets'
      ...
      Plug 'vimwiki/vimwiki', { 'branch': 'dev' }
      call plug#end()

- [x] Post the syntax you are using (default/mediawiki/markdown) **and** your vimwiki settings from your `.vimrc`
- Using 'default' vimwiki syntax NOT 'mediawiki' or 'markdown'
- My vimviki settings are as follows in my `.vimrc`:
      
      hi VimwikiHeader1 ctermfg=Red guifg=#FF0000
      hi VimwikiHeader2 ctermfg=Green guifg=#00FF00
      hi VimwikiHeader3 ctermfg=Blue guifg=#0000FF
      hi VimwikiHeader4 ctermfg=Magenta guifg=#FF00FF
      hi VimwikiHeader5 ctermfg=Cyan guifg=#00FFFF
      hi VimwikiHeader6 ctermfg=Yellow guifg=#FFFF00

- [x] Provide a detailed description of the problem including **steps to reproduce the issue**
- Issue Description: 

1. After installation of vimwiki plugin in files having types of '.wiki' `<Tab>`, `<S-Tab>` and `<CR>` keys won't work as intended by `coc-nvim` plugin in the `Insert Mode` of vim. `<Tab>` and `<S-Tab>` keys insert a `space` character by instantly closing popmenu without any selection instead of selecting a snippet to expand and `<CR>` key insert a `newline` character instead of selecting and inserting the 1st snippet

2. Out of `Insert Mode` of vim `<Tab>`, `<S-Tab>` cycles through the links and '<CR>' creates/follows links in the `.wiki` file as instended by `vimwiki` plugin.

3. Suggested methods defined in issue #1048 and #845 didn't work in my circumstances.
4. Method suggested by madskjeldgaard in issue #845 only helps to insert the abbreviation of the snippet displayed in popmenu in the `Insert Mode`, related snippet never expands. It only allows you to select the related snippet abbreviation

Conclusion: Could we use the related keys in `.wiki` files to insert snippets by using the behaviour defined by `coc-nvim` plugin's "Example Vim-Configuration" as well as to navigate through `.wiki` files by using the behaviour defined by 'vimwiki' plugin ???
 
- Issue Reproduction Steps For `master` branch and `dev` branch

1. Add the plugin related lines to your `.vimrc` shown in the first section of this issue submission
2. Install plugins by executing `:PlugInstall' command in `vim`
3. Bu using `:CocConfig` enter the following line to the coc configuration file located at `~/.vim/coc-settings.json` for using only the snipMate snippets:
```
    "snippets.ultisnips.enable": false,
```
4. Add "Example Vim-Configuration" from [here](https://github.com/neoclide/coc.nvim) with the following additional line recommended to be used
      for Slackware 14.2 to the `.vimrc`
```
" Use tab for trigger completion with characters ahead and navigate.
" NOTE: Use command ':verbose imap <tab>' to make sure tab is not mapped by
" other plugin before putting this into your config.
inoremap <silent><expr> <TAB>
      \ pumvisible() ? "\<C-n>" :
      \ <SID>check_back_space() ? "\<TAB>" :
      \ coc#refresh()
execute 'set t_kB=' . nr2char(27) . '[Z'   <- This is the line to be added to use <S-Tab> in vim under Slackwarer 14.2
inoremap <expr><S-TAB> pumvisible() ? "\<C-p>" : "\<C-h>"
``` 
5. Open `index.wiki` located in `~/vimwiki/` by using `<Leader>ww` keymap in `vim`,
      Enter character 'G' in line 12 of the `index.wiki' file (pop-menu has opened as expected to show
      related snippet abbreviations to insert by expanding)
  ![Vimwiki issue master branch screenshot1](https://i.ibb.co/VWKpZFQ/master1.png)
6. Hit `<Tab>` in `Insert Mode` of vim (instead of selecting 'GPL2~' snippet from the list for expansion
      inserts a 'space' character after the character 'G')
                                    OR
      Hit `<S-Tab>` in `Insert Mode` of vim (instead of selecting 'AGPL3~' snippet from the list for expansion
      inserts a 'space' character after the character 'G')
![Vimwiki issue master branch screenshot2](https://i.ibb.co/VSJCHWF/master2.png)
7. Hit `<CR>` immediately after step 3 without hitting `<Tab>` or `<S-Tab>` as defined in step 4 in `Insert Mode'
      of vim (instead of selecting 'GPL2~' snippet and inserting the expansion jumps to the newline located at line 13)
![Vimwiki issue master branch screenshot3](https://i.ibb.co/n3fZs0c/master3.png)

- [x] Include the output of `:VimwikiShowVersion`.
- For `master` branch output of `:VimwikiShowVersion` is as follows:

      Stable Version: 2.5

- For `dev` branch output of `:VimwikiShowVersion` is as follows

      OS: Linux
      Vim: 802
      Branch: dev
      Revision: a9d157f
      Date: 2021-01-15  13:48:13  -0300

I'd like to thank anybody who has read up to this point for his/her patience and efforts for maintaining this very useful plugin.

**Comment by tinmarino** [comment]
Hi @N-Tek 

I understand you want tab mapping from coc and not from vimwiki in case of conflict.

Can you try to load first vimwiki (before coc) in your .vimrc?

```vim
call plug#begin('~/.vim/plugged')
...
Plug 'vimwiki/vimwiki', { 'branch': 'dev' }
let g:coc_global_extensions = ['coc-bookmark',
                                           \ 'coc-calc',
  	                             \ 'coc-clangd',
  	                             \ 'coc-cmake',
  	                             \ 'coc-css',
  	                             \ 'coc-cssmodules',
  	                             \ 'coc-git',
  	                             \ 'coc-highlight',
  	                             \ 'coc-html',
  	                             \ 'coc-json',
  	                             \ 'coc-markdownlint',
  	                             \ 'coc-r-lsp',
  	                             \ 'coc-sh',
  	                             \ 'coc-snippets',
  	                             \ 'coc-stylelint',
  	                             \ 'coc-stylelintplus',
  	                             \ 'coc-svg',
  	                             \ 'coc-template',
  	                             \ 'coc-todolist',
  	                             \ 'coc-tsserver',
  	                             \ 'coc-vimlsp',
  	                             \ 'coc-xml']
Plug 'neoclide/coc.nvim', {'branch': 'release'}
```

Also I advice you to configure coc __before__ loading it. For coc, it may not change anything but for vimwiki, variables are read at initialization (which is a bad design but must be followed now).

**Comment by N-Tek** [comment]
Hi @tinmarino,

First of all I'd like to thank you for your immediate reply :D

Yes, I want to use coc mappings instead of the vimwiki ones in case of a conflict to insert snippets into `.wiki` files.

I've tried the method you've suggested by first uninstalling both plugins and then reinstalling them in the suggested order according to your directives but unfortunately the behaviour of 
`<Tab>`, `<S-Tab>` and `<CR>` hasn't changed, still `<Tab>` and `<S-Tab>` inserts a 'space' character and `<CR>` hops on the line below the current one in `Insert Mode` of vim and navigate through or create vimwiki links in `Normal Mode` of vim.

I'm not sure whether the following info will ease the problem or not but I'd like to share the following:

**OS:**  Slackware 14.2 with XFCE 4.12
**Locale:** tr_TR.utf8
**Terminal Emulator:** xfce4-terminal 0.6.3 with tmux 2.1
                                    (TERM is set in .tmux.conf as "screen-256colors")

- mappings in "index.wiki" are as follows:
![mappings](https://i.ibb.co/Ch4LP4v/mappings.png)

Also your advices are much appreciated, I think they're gold for self-learners like myself :D

**Comment by cbows** [comment]
I am having the same issue without coc. I am using ncm2 with neovim nightly. Changing the order also didn't work.

**Comment by mtoohey31** [comment]
I believe I've found a solution that may work for this, I was having a similar issue with coc's autocomplete (not snippets specifically though, I don't use them right now), where when pressing the enter key, the vimwiki insert mode mapping would take precedence for `<CR>` over selecting the hovered autocomplete option. I solved this by handling my mappings manually, and by using the `CompleteChanged` and `CompleteDone` autocommands:
```vim
autocmd FileType markdown,rmarkdown,vimwiki " activates only on markdown, rmarkdown, and vimiwki filetypes
      \ inoremap <CR> <ESC>:VimwikiReturn 3 5<CR>| " enables insert mode vimwiki return mapping
      \ autocmd CompleteChanged * silent! iunmap <CR>| " disables insert mode mapping on complete changed
      \ autocmd CompleteDone * if ! empty(v:completed_item) | inoremap <CR> <ESC>:VimwikiReturn 3 5<CR>| endif " re-enables the insert mode mapping
```
This lets you keep using the vimwiki insert mode mapping, but allows you to use enter for selecting an autocomplete option while in insert mode.
I haven't tested this too much yet, but it seems to be working well so far, I'll update this if that changes.

**Comment by N-Tek** [comment]
Unfortunately both disabling table_mappings and adding autocommands didn't work for me but I'd like to thank everybody  for sharing their experiences

**Comment by MagicDuck** [comment]
@N-Tek I also ran into this problem while using `nvim-compe` completion plugin and `vimwiki`
It seems to have to do with this insert mode mapping in the `vimwiki` code:
https://github.com/vimwiki/vimwiki/blob/619f04f89861c58e5a6415a4f83847752928252d/ftplugin/vimwiki.vim#L487
What worked for me was to create a file `after/ftplugin/vimwiki.vim` with the following contents:
```vim
inoremap <silent><buffer><expr> <CR>      compe#confirm('<C-]><Esc>:VimwikiReturn 1 5<CR>')
```
Note that this mapping is as per the `nvim-compe` github page with 2 additions:
1. added `<buffer>`
2. the parameter passed to `compe#confirm('<stuff>')` is the right hand side of the mapping from:
https://github.com/vimwiki/vimwiki/blob/619f04f89861c58e5a6415a4f83847752928252d/ftplugin/vimwiki.vim#L487

I think the fix needs to depend on whatever completion plugin you are using, which is why I am guessing @mtoohey31 was successful as well.

**Comment by N-Tek** [comment]
Hi MagicDuck as far as I understand you're using a totally different plugin for auto completion but you've shared your way that have worked for yourself to help which I appreciated a lot. I'd like to report what I've done as follows:

1) created a folder with the following command
    `mkdir ~/.vim/after/ftplugin && cd ~/.vim/after/ftplugin`

2) pasted line 552 of my `~/.vim/plugged/vimwiki/ftplugin/vimwiki.vim` file to `~/.vim/after/ftplugin/vimwiki.vim`

    Near vicinity of line 552 is as follows: 

                      call vimwiki#u#map_key('n', 'o', '<Plug>VimwikiListo')
                      call vimwiki#u#map_key('n', 'O', '<Plug>VimwikiListO')

            545   " Set lists_return to 0, if you don't want <CR> mapped to VimwikiReturn
                     if str2nr(vimwiki#vars#get_global('key_mappings').lists_return)
                     " Handle case of existing VimwikiReturn mappings outside the <Plug> definition
                     " Note: Avoid interfering with popup/completion menu if it's active (#813)
                        if maparg('<CR>', 'i') !~# '.*VimwikiReturn*.'
                           if has('patch-7.3.489')
                           " expand iabbrev on enter
             552 ->             inoremap <expr><silent><buffer> <CR> pumvisible() ? '<CR>' : '<C-]><Esc>:VimwikiReturn 1 5<CR>'
                           else
                                inoremap <expr><silent><buffer> <CR> pumvisible() ? '<CR>' : '<Esc>:VimwikiReturn 1 5<CR>'
                           endif
                        endif
                        if  maparg('<S-CR>', 'i') !~# '.*VimwikiReturn*.'
                              noremap <expr><silent><buffer> <S-CR> pumvisible() ? '<CR>' : '<Esc>:VimwikiReturn 2 2<CR>'
                        endif
                    endif
  
                    " change symbol for bulleted lists
                    for s:char in vimwiki#vars#get_syntaxlocal('bullet_types')
             564    if !hasmapto(':VimwikiChangeSymbolTo '.s:char.'<CR>')`
 

  

3) Quitted vim and restart it again, opened a vimwiki page, unfortunately nothing has changed, behaviour I've mentioned remains the same
     
I've also tried to add first line 554 individually, nothing has changed,
then each line one-by-one starting from 545 that is in if-else branching block up to line 560 again nothing has changed,
and finally whole nested if-else branching block between line 545 & 560 nothing has changed

During all those experiments I haven't touched anything in my `~/.vimrc` file

But I'd like to thank you again for sharing your knowledge :D

**Comment by anton-fomin** [comment]
I had the same issue and the following helped me

```
" disable table mappings
let g:vimwiki_key_mappings = {
            \ 'all_maps': 1,
            \ 'global': 1,
            \ 'headers': 1,
            \ 'text_objs': 1,
            \ 'table_format': 1,
            \ 'table_mappings': 0,
            \ 'lists': 1,
            \ 'links': 1,
            \ 'html': 1,
            \ 'mouse': 0,
            \ }
augroup VimwikiRemaps
    autocmd!
    " unmap tab in insert mode
    autocmd Filetype vimwiki silent! iunmap <buffer> <Tab>
    " remap table tab mappings to M-n M-p
    autocmd Filetype vimwiki inoremap <silent><expr><buffer> <M-n> vimwiki#tbl#kbd_tab()
    autocmd Filetype vimwiki inoremap <silent><expr><buffer> <M-p> vimwiki#tbl#kbd_shift_tab()
    " on enter if completion is open, complete first element otherwise use
    " default vimwiki mapping
    autocmd Filetype vimwiki inoremap <silent><expr><buffer> <cr> pumvisible() ? coc#_select_confirm()
                              \: "<C-]><Esc>:VimwikiReturn 1 5<CR>"
augroup end

```

**Comment by shafouz** [comment]
@anton-fomin thanks for the solution it works perfectly

**Comment by bbggl** [comment]
@anton-fomin thanks for the solution it works perfectly

**Comment by reportaman** [comment]
@anton-fomin Thanks so much! It works well.
